cat/spirv.cat

```diff
@@ -0,0 +1,127 @@
+SPIRV
+
+(* Terms from spirv.als *)
+// stor = [] "set to relation, (s,s) for all s in E, lift a set to a relation (the [_] operator in Herd)"
+// rai-r = r | r^-1
+// sthd = int
+// immpo = po - po;po
+// rc[r] = r?
+// ssg, swg, sqf = same scope id @ subgroup, workgroup, queuefamily
+// inscope = sr
+// sloc = loc
+// sref = vloc
+// scbarinst = syncbar "relates pairs of control barriers, same control barrier dynamic instance"
+// imm-r = r - (r;r+)
+// twoplus-r = r;r+
+
+//???????????
+// chains = new relation
+// ssw = new relation for system-synchronizes-with?
+
+(*******************)
+(*  Auxiliaries *)
+(*******************)
+let mutordatom = [ATOM]; ((loc & vloc & sr) \ id); [ATOM] // mutually ordered atomics (same loc, inscope)
+
+let sctosem = SC0 * SEMSC0 | SC1 * SEMSC1
+let semtosc = SEMSC0 * SC0 | SEMSC1 * SC1
+
+let avvisSem = ((SC0 | SC1) * ((AV | VIS) & DV)) | (sctosem; [SEMAV | SEMVIS])
+let avvisSet = [AV | VIS]; (vloc & loc)
+let avvisinc = (avvisSem | avvisSem^-1) \ id |
+               (avvisSet | avvisSet^-1)
+
+let posctosem = sctosem & po
+let posemtosc = semtosc & po
+
+let asmo = [ATOM & W]; mutordatom; [ATOM & W]
+let rs = [REL & ATOM]; ((asmo \ (asmo; asmo+)); [R & W])*
+let hypors = [W & ATOM]; ((asmo \ (asmo; asmo+)); [R & W])*
+
+// synchronizes-with is similar to C++, with an additional case for fence->cbar->cbar->fence
+let sw = sr & (
+    ([REL & ATOM]; rs; (rf & mutordatom); [ACQ & ATOM]) |                                               // atomic->atomic
+    ([REL & F]; posemtosc; [ATOM & W]; hypors; (rf & mutordatom); [ACQ & ATOM]) |                       // fence->atomic
+    ([REL & ATOM]; rs; (rf & mutordatom); [ATOM & R]; posctosem; [ACQ & F]) |                           // atomic->fence
+    ([REL & F]; posemtosc; [ATOM & W]; hypors; (rf & mutordatom); [ATOM & R]; posctosem; [ACQ & F]) |   // fence->fence
+    ([REL & F]; po?; [CBAR]; (syncbar & sr \ id); [CBAR]; po?; [ACQ & F])                               // fence->cbar->cbar->fence,
+    // (stor[CBAR]) terms are redundant because scbarinst is an equivalence relation on scbarinst, but they make the sequence of instructions more clear.
+)
+
+let ithbsemsc0 = ( ssw |
+                   ([SEMSC0]; sw; [SEMSC0]) |
+                   ([SC0 | SEMSC0]; po; [REL & SEMSC0]) |
+                   ([ACQ & SEMSC0]; po; [SC0 | SEMSC0])
+                 )+
+let ithbsemsc1 = ( ssw |
+                   ([SEMSC1]; sw; [SEMSC1]) |
+                   ([SC1 | SEMSC1]; po; [REL & SEMSC1]) |
+                   ([ACQ & SEMSC1]; po; [SC1 | SEMSC1])
+                 )+
+let ithbsemsc01 = ( ssw |
+                   ([SEMSC01]; sw; [SEMSC01]) |
+                   ([SC0 | SC1 | SEMSC01]; po; [REL & SEMSC01]) |
+                   ([ACQ & SEMSC01]; po; [SC0 | SC1 | SEMSC01])
+                 )+
+
+let hb = ithbsemsc0 | ithbsemsc1 | ithbsemsc01 | po // happens-before
+
+//let chains = (M | F | CBAR) * (M | F | CBAR)
```

Comment:
As we discuss, uncomment this one and add the `chains` to everywhere where it is used. Add a comment saying this is unsound for those litmus tests that had the `NOCHAINS` at the end

Reply:
Updated